pilot/pkg/model/push_context.go

```diff
@@ -259,18 +259,34 @@ func (first *PushRequest) Merge(other *PushRequest) *PushRequest {
 		Reason: append(first.Reason, other.Reason...),
 	}
 
-	// Only merge EdsUpdates when incremental eds push needed.
-	if !merged.Full {
-		merged.EdsUpdates = make(map[string]struct{})
-		// Merge the updates
-		for update := range first.EdsUpdates {
-			merged.EdsUpdates[update] = struct{}{}
+	// Do not merge when any one is empty
+	if len(first.ConfigsUpdated) > 0 && len(other.ConfigsUpdated) > 0 {
+		merged.ConfigsUpdated = make(map[resource.GroupVersionKind]map[string]struct{})
+		for kind := range first.ConfigsUpdated {
+			merged.ConfigsUpdated[kind] = make(map[string]struct{})
 		}
-		for update := range other.EdsUpdates {
-			merged.EdsUpdates[update] = struct{}{}
+		for kind := range other.ConfigsUpdated {
+			merged.ConfigsUpdated[kind] = make(map[string]struct{})
+		}
+
+		if !merged.Full {
+			for kind := range merged.ConfigsUpdated {
+				d1 := first.ConfigsUpdated[kind]
+				d2 := other.ConfigsUpdated[kind]
+
+				for update := range d1 {
+					if merged.ConfigsUpdated[kind] != nil {
```

Comment:
No need to check nil, guaranteed by above line 266 and 269

Reply:
okay, nice catch. done